### PR TITLE
[CEN-1270] REST interface for CSV files upload

### PR DESCRIPTION
**Pull request by alberto-re-pagopa**
This PR implements an alternative to SFTP for the upload of transaction files produced by acquirers. It leverages the existing pagopastorage Blob Storage and uses Service Shared Access Signatures to grant scoped, temporary, upload permissions to authenticated clients.

### List of changes

Implemented a new APIM API 'RTD CSV Transaction' which exposes two methods granting write permissions for CSTAR and AdE transaction files.

### Motivation and context

This change allows us to not be more dependent of externally managed FTP servers for transactions's uploads. It also makes more easy to trigger custom code on upload events.

### Type of changes

- [x] Add new resources
- [ ] Update configuration to existing resources
- [ ] Remove existing resources

### Does this introduce a change to production resources with possible user impact?

- [ ] Yes, users may be impacted applying this change
- [x] No

### Does this introduce an unwanted change on infrastructure? Check terraform plan execution result

- [ ] Yes
- [x] No

### Other information

---

### CODE REVIEW

**Review by pasqualedevita — APPROVED** [review]
I trust in u 👍 

please put in cstar backlog tfsec static analysis suggestions https://github.com/pagopa/cstar-infrastructure/runs/4868457877?check_suite_focus=true (probably they are already in the backlog but double check it)

File: src/api.tf [hunk 1]
```diff
@@ -432,6 +432,52 @@ module "rtd_payment_instrument_manager" {
   ]
 }
 
+## RTD CSV Transaction API ##
+module "rtd_csv_transaction" {
+  source = "git::https://github.com/pagopa/azurerm.git//api_management_api?ref=v2.0.28"
```
Comment (GiovanniMancini) [inline src/api.tf:437]: Maybe we could upgrade to the latest version of `azurerm`

File: src/api.tf [hunk 2]
```diff
@@ -432,6 +432,52 @@ module "rtd_payment_instrument_manager" {
...
+  api_management_name = module.apim.name
+  resource_group_name = azurerm_resource_group.rg_api.name
+
+
+  description  = "API providing upload methods for csv transaction files"
+  display_name = "RTD CSV Transaction API"
+  path         = "rtd/csv-transaction"
+  protocols    = ["https", "http"]
+
+  service_url = format("https://%s", module.cstarblobstorage.primary_blob_host)
```
Comment (GiovanniMancini) [inline src/api.tf:449]: This variable should reference a private url

File: src/api/rtd_csv_transaction/create-sas-token-policy.xml.tpl [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+            }"
+        />
+        <!--
+            END - Create a Shared Key to authorize APIM to create a container into a storage account
+        -->
+
+        <!--
+            START - Perform a PUT operation on the storage REST endpoint
+        -->
+        <set-variable name="createContainerUrl" value="@{
```
Comment (GiovanniMancini) [inline src/api/rtd_csv_transaction/create-sas-token-policy.xml.tpl:121]: The storage should be accessed via a private url (e.g. a private endpoint), maybe we could fix it as a part of another pr

File: src/api/rtd_csv_transaction/create-sas-token-policy.xml.tpl [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+                return "https://" + context.Variables["storageAccount"] + ".blob.core.windows.net/" + context.Variables["containerName"] + "?restype=container";
+            }"
+        />
+
+        <send-request mode="new" response-variable-name="createContainerResult" timeout="20" ignore-error="false">
```
Comment (GiovanniMancini) [inline src/api/rtd_csv_transaction/create-sas-token-policy.xml.tpl:126]: Could we use a lower value?

**Comment by GiovanniMancini** [comment]
> I trust in u 👍
> 
> please put in cstar backlog tfsec static analysis suggestions https://github.com/pagopa/cstar-infrastructure/runs/4868457877?check_suite_focus=true (probably they are already in the backlog but double check it)

Thank you. We have expanded the backlog to implement many changes suggested by `tfsec`.. the most important is the hardening of the blob storage, as we are starting to use it to exchange file via http, thus avoiding ftp

**Comment by pasqualedevita** [comment]
> > I trust in u 👍
> > please put in cstar backlog tfsec static analysis suggestions https://github.com/pagopa/cstar-infrastructure/runs/4868457877?check_suite_focus=true (probably they are already in the backlog but double check it)
> 
> Thank you. We have expanded the backlog to implement many changes suggested by `tfsec`.. the most important is the hardening of the blob storage, as we are starting to use it to exchange file via http, thus avoiding ftp

for storage account we can use private endpoints. They are standard in our infrastructure and simple to set.
If u need I can help you

**Review by GiovanniMancini — APPROVED** [review]

**Pull request merged**